sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,492 @@
+    private final long batchSize;
+
+    private Write(DataSourceConfiguration dataSourceConfiguration, String statement,
+                  PreparedStatementSetter<T> preparedStatementSetter, long batchSize) {
+      this.dataSourceConfiguration = dataSourceConfiguration;
+      this.statement = statement;
+      this.preparedStatementSetter = preparedStatementSetter;
+      this.batchSize = batchSize;
+    }
+
+    @Override
+    public PDone apply(PCollection<T> input) {
+      input.apply(
+          ParDo.of(new WriteFn<T>(dataSourceConfiguration, statement, preparedStatementSetter,
+              batchSize)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<T> input) {
+      dataSourceConfiguration.validate();
+      Preconditions.checkNotNull("statement", statement);
+      Preconditions.checkNotNull("preparedStatementSetter", preparedStatementSetter);
+    }
+
+    private static class WriteFn<T> extends DoFn<T, Void> {
+
+      private final DataSourceConfiguration dataSourceConfiguration;
+      private final String statement;
+      private final PreparedStatementSetter<T> preparedStatementSetter;
+      private long batchSize;
+
+      private DataSource dataSource;
+      private Connection connection;
+      private PreparedStatement preparedStatement;
+      private long batchCount;
+
+      public WriteFn(DataSourceConfiguration dataSourceConfiguration, String statement,
+                     PreparedStatementSetter<T> preparedStatementSetter, long batchSize) {
+        this.dataSourceConfiguration = dataSourceConfiguration;
+        this.statement = statement;
+        this.preparedStatementSetter = preparedStatementSetter;
+        this.batchSize = batchSize;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        dataSource = dataSourceConfiguration.getDataSource();
+        if (dataSourceConfiguration.username != null) {
+          connection = dataSource.getConnection(dataSourceConfiguration.username,
+              dataSourceConfiguration.password);
+        } else {
+          connection = dataSource.getConnection();
+        }
+        connection.setAutoCommit(false);
+        preparedStatement = connection.prepareStatement(statement);
+      }
+
+      @StartBundle
+      public void startBundle(Context context) {
+        batchCount = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        T record = context.element();
+
+        preparedStatement.clearParameters();
+        preparedStatementSetter.setParameters(record, preparedStatement);
+        preparedStatement.addBatch();
+
+        batchCount++;
+
+        if (batchCount >= batchSize) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        preparedStatement.executeBatch();
```

Comment:
Can you check that this doesn't throw an exception in case the bundle is empty? (empty bundles are rare but they exist)